### PR TITLE
Feature/feature 1

### PR DESCRIPTION
**Pull request by MiguelArgentina**
#Milestone: Advanced Building Blocks - Enumerables

Project setup & assets
Set up repository, files and linters.

Features added
implemented our own version of Ruby’s enumerable methods.
Also, we are complying with Git Flow requirements, working in a Feature branch and leaving the develop and main branches for clean merges only.

Documentation
Created the read me file, license file, checked for linter errors and created a pull request.

### CODE REVIEW

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGE REQUIRED!! :recycle: :red_circle: 
-----------------------
Hello team!! :wave: you are doing a great job :clap: :1st_place_medal: 
There are some errors to fix :hammer_and_wrench: see them below :point_down: 
----------------

- [x] Your linters are failing, before request a review check that the linters are passing.
- [x] Set up your `development` branch as a default branch. The pull request should be `feature/feature_1`  to `development` not to `main` branch. [Here](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/) you can find the Git Flow guidelines.
- [x] Put a question mark on `my_all`, `my_any` and `my_none`, so it will be like this `my_all?`, `my_any?`, `my_none?`, those are the names that are required on this project. 

- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self. Is returning all the numbers on the array E.g [5, 6, 7, 8, 9 ..... 49, 50] and it should return 5..50.
- [x] #my_inject raises a "LocalJumpError" when no block or argument is given. Instead nothing was raised. 
-------------------
Make the required changes and ask for another review :muscle: :+1: 
HAPPY CODDING!! 

MB5W4D

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :recycle: :red_circle: 
------------------------
Hello team :wave: excellent job working on the required changes :muscle: :+1: 
There are a few changes to fully complete the project :hammer_and_wrench: see them below :point_down: 
---------------

- [x] Remember to set up in GitHub your `development` branch as the default branch. 
- [x] You change the name of your method `my_any?` but you forget to change it where you call it. In `my_none?` in lines 70 and 72 changes the name, which means just put the question mark `?` at the end of `my_any`
- [x] In line 17 changes `length` to `size`, This `length` inbuilt method causing troubles with a range. 
-----------------
With the request changes made :point_up: the project should work perfectly :muscle: :1st_place_medal:. once you are done send it for another review :smile: 
HAPPY CODING!!

**Comment by MiguelArgentina** [comment]
Thanks for your comments!

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    false
+  end
+
+  def my_none?(param = nil)
+    if block_given?
+      !my_any?(&Proc.new)
+    else
+      !my_any?(param)
+    end
+  end
```
Comment (enaburekhan) [inline enumerables.rb]: enumerables #my_none? does not mutate the original array
     Failure/Error: array.my_none? { |num| num + 1 }
     
     ArgumentError:
       tried to create Proc object without a block
     # ./enumerables.rb:70:in `new'
     # ./enumerables.rb:70:in `my_none?'
     
Currently,  RSpec is flagging the .new in line 70.
You need to rewrite my_none? method.  Use the already created my_each method to loop through the array like this:
def my_none?(parameter=nil)
  my_each do |y|....check all the conditions using if/elsif  inside the block_given?...end true end

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes requested:full_moon:
**Hi team!
Great job on implementing some of the changes requested by the previous Tse's:clap::clap:. However, there are still a few changes we need to make to get all the tests to pass and the project approved :muscle:.**

- [x] We are getting all four of these errors from my_none? : **"tried to create Proc object without a block"**. Kindly check my comment below on the  PR codebase on how to solve the issue and get all your test to pass. see the errors screen-shot for reference: https://imgur.com/a/PbciGcA

Make the above-required changes and when you are done, request for another review.
You can drop your questions :point_down: if you have any and ping me on slack @ Eric Enaburekhan to respond.

Happy Coding :blush:
Eric Enaburehan

File: enumerables.rb (hunk 1 above)
Comment (MiguelArgentina) [reply enumerables.rb]: Thanks! Really appreciate the explanation!

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Request changes :full_moon:
**Great job so far on implementing almost all the requested changes:clap::clap:. However, one of the tests is still not passing. Let's fix this so we can get this project approved :muscle:.**

- [ ] enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex.
 Currently, the method is returning true instead of false. You need to also test for Regex in #my_none?. Fix this to get all the tests to pass.

Make the above-required changes and when you are done, request for another review
Kindly drop your questions:point_down: if you have any and ping me on slack @ Eric Enaburehan to respond.
Happy Coding:blush:
Eric Enaburekhan

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    when Class
+      to_a.my_each { |item| return false unless [item.class, item.class.superclass].include?(param) }
+    when Regexp
+      to_a.my_each { |item| return false unless param.match(item) }
+    else
+      to_a.my_each { |item| return false if item != param }
+    end; true
+  end
+
+  def my_any?(param = nil)
```
Comment (GzimAsani) [inline enumerables.rb:46]: 1) enumerables #my_any? returns false if the block ever returns a value false or nil
      Failure/Error: expect(array.my_any?(&false_block)).to eq(array.any?(&false_block))
 2) enumerables #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
3) enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true
      Failure/Error: expect(false_array.my_any?).to be false_array.any?

  4) enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
5) enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
6) enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
      Failure/Error: expect(words.my_any?(/d/)).to be words.any?(/d/)
 7) enumerables #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
      Failure/Error: expect(words.my_any?(/z/)).to be words.any?(/z/)
 8) enumerables #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
9) enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

**Keep in mind about my_any? that: The method returns true if the block ever returns a value other than false or nil. If the block is not given, Ruby adds an implicit block of { |obj| obj } that will cause any? to return true if at least one of the collection members is not false or nil.**

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  def my_select
+    return to_enum unless block_given?
+
+    ary_aux = []
+    my_each { |item| ary_aux << item if yield item }
+    ary_aux
+  end
+
+  def my_all?(param = nil)
```
Comment (GzimAsani) [inline enumerables.rb:30]: 1) enumerables #my_all? returns false if the block returns false or nil
     Failure/Error: expect(array.my_all?(&false_block)).to eq(array.all?(&false_block))
 2) enumerables #my_all? returns false if the block returns false or nil::range
     Failure/Error: expect(range.my_all?(&false_block)).to eq(range.all?(&false_block))

  3) enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
 4) enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
5) enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
6) enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
 7) enumerables #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
 8) enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
     Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)
 9) enumerables #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
  10) enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
 11) enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

**Keep in mind abut my_all? that  The method returns true if the block never returns false or nil. If the block is not given, Ruby adds an implicit block of { |obj| obj } which will cause all? to return true when none of the collection members are false or nil.**

**Review by GzimAsani — CHANGES_REQUESTED** [review]
###  Status: Changes Required :recycle:
**Hello team** very nice job implementing all the changes from the previous reviews, however we still have some issues to fix, kindly check the comments below for more detailed information:
- [ ] You have to improve your my_any? and my_all? because they are throwing errors, I would suggest that you rewrite them.
- _[ HINT]_ do not get discouraged by all the comments because out of all your methods only 2 are throwing errors and you have done great job until now.

**Review by mke2111 — APPROVED** [review]
# **🎉 🎉 STATUS: APPROVED 🎉 🎉** 

### Congratulations Team. 🥇 

**You finally made it to the 'success lounge', You should reflect on everything you have sacrificed to get to this point in the program and life. **Kudos to you. 🍹**** 

Keep the candle lighting and going for more wins.

### You may now merge to the develop branch and submit a project completion form.

### Cheers!

### Reach me on slack for a chitchat 🆘  **Roy Mukuye**

**Comment by MiguelArgentina** [comment]
Great! Thanks to all the TSEs for your reviews!

**Pull request merged**